add arrow key focus to distance expense

jsdev2547 opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ https://github.com/Expensify/App/issues/34350
PROPOSAL: https://github.com/Expensify/App/issues/34350#issuecomment-1971784283


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- [x] Verify that no errors appear in the JS console

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

1. Create distance expense.
2. Enter Start address to display some results.
3. Verify the search results are focusable using arrow keys and can be selected using enter key.
4. Clear results and verify current location item and recent destinations are focusable and can be selected using enter key.

- [x] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->
N/A
</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->
N/A
</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->
N/A
</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->
N/A
</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->

[web.webm](https://github.com/Expensify/App/assets/161593779/98efb8b5-6a8c-4455-a153-e003d01799da)

</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->

[desktop.webm](https://github.com/Expensify/App/assets/161593779/236ae4c4-b869-4a97-b3fd-6035c67d0ec8)

</details>

jsdev2547 commented:
@abdulrahuman5196 I have resolved the conflicts. PR is ready for review.

abdulrahuman5196 commented:
Hi, Starting to review today only given the team's decision. @jsdev2547 could you kindly fix the conflicts?

And is the old PR removed? I don't see it to check on old comments?

jsdev2547 commented:
@abdulrahuman5196 PR is ready for review. I do not have access to previous PR and comments.

jsdev2547 commented:
@abdulrahuman5196 I do not have access to previous PR and comments. PR is ready for review.

jsdev2547 commented:
@abdulrahuman5196 Please finish PR review.

abdulrahuman5196 commented:
Hi, @jsdev2547  sorry the delay. Will review in couple of hours.

abdulrahuman5196 commented:
@jsdev2547 The author's checklist doesn't have tests and there is no video recordings present?

abdulrahuman5196 commented:
@jsdev2547 And I am still seeing issues with current location arrow selection

https://github.com/Expensify/App/assets/46707890/7b815f16-6910-41eb-9be8-d7d2853876c6

roryabraham commented on src/components/AddressSearch/index.tsx:
```diff
@@ -407,14 +450,39 @@ function AddressSearch(
+                                if (!e) {
+                                    return;
+                                }
+                                if (e.code === 'ArrowUp') {
```
Why do you need to manage the arrow keys outside of `useArrowKeyFocusManager`?

jsdev2547 replied:
@roryabraham The arrowkeyfocusmanager can only focus on search result items but the current location item is not included in the search results.

abdulrahuman5196 replied:
@jsdev2547 I too agree with the question above.

We already have,

```
const [focusedIndex, setFocusedIndex] = useArrowKeyFocusManager({
        initialFocusedIndex: -1,
        maxIndex: maxIndex - 1,
        isActive: true,
    });
```

`focusedIndex` here represent the index including the currentLocation header and the search list. 
Why can't we have it like, when the focusIndex == 0, focus currentLocation button, and the rest for the suggestion list. 
We can have the `maxIndex as suggestionCount + currentLocationItem` when we show currentLocation. (Throwing idea on high level.)

I am not sure of the need by we need to listen to 'ArrowUp' here also, which is what `useArrowKeyFocusManager` does and returns us the focusedIndex.

jsdev2547 replied:
@abdulrahuman5196 I tried to use onFocusedIndexChange callback of useArrowKeyFocusManager instead of the onKeyPress event but the arrowkeyfocusmanager has reserved index 0 for search list and the focusindex -1 is when the search list is empty, so even if maxIndex is set to 1 to cover for currentLocationItem, the focusIndex would reset to 0 when the search is entered and focusIndex 0 would be for the search list. The arrowkeyfocusmanager has some flexibility and can leverage onFocusedIndexChange but the difficulty is presented by the current location item which is not part of the search list.

jsdev2547 replied:
@roryabraham @abdulrahuman5196 I suggest to remove the ArrowKeyFocusManager and use KeyPress event instead. I have tested using KeyPress event and the ArrowKeyFocusManager is not needed. The ArrowKeyFocusManager only focuses if the search result is not empty but the current location item needs focus when the search list is empty.

abdulrahuman5196 replied:
IMO ArrowKeyFocusManager is just a effective wrapper to handle the key press. It would be better to use it or update it if required.
@jsdev2547 Could you kindly provide the code pointers for the above concern/analysis? Let me also see if there is something else we can do?

jsdev2547 replied:
@abdulrahuman5196 Please get latest and review.

jsdev2547 replied:
@abdulrahuman5196 I have pushed the revised KeyPress event handler. Please get latest and review.

jsdev2547 replied:
@abdulrahuman5196 I have pushed the revised KeyPress event handler. Please review.

jsdev2547 commented:
@abdulrahuman5196 I have added test steps and also resolved the maxIndex not getting reset. PR is ready for review.

jsdev2547 commented:
@abdulrahuman5196 I have added test steps. The maxIndex needed to reset after search results are cleared. I have resolved the focusing issue. PR is ready for review.

jsdev2547 commented:
@abdulrahuman5196 I have added videos.

abdulrahuman5196 commented:
Thank you for the videos. I will work on platform testing in my morning.

abdulrahuman5196 commented:
Reviewing now

jsdev2547 commented:
@abdulrahuman5196 Thanks for testing. When no search results are present, the early return statement would return and the current location item button press would not get handled. I have resolved the issue.

jsdev2547 commented:
@abdulrahuman5196 I have revised the platform testing videos.

abdulrahuman5196 commented:
Reviewing again

abdulrahuman5196 commented:
> IMO ArrowKeyFocusManager is just a effective wrapper to handle the key press. It would be better to use it or update it if required. @jsdev2547 Could you kindly provide the code pointers for the above concern/analysis? Let me also see if there is something else we can do?

@jsdev2547 I am not sure why we have removed the `useArrowKeyFocusManager` and changed to raw key press handling? The suggestion was to investigate if we can improve the `useArrowKeyFocusManager` if required.
Could you kindly respond to the above query?

jsdev2547 commented:
@abdulrahuman5196 The useArrowKeyFocusManager returns early if the maxIndex is less than 0 and resets index to -1 when no search results, which is why I had previously used arrowkeyfocusmanager alongside with keypress event to handle currentlocation item selection.
https://github.com/Expensify/App/blob/c848cb6bf29b600eecb8b338222fffe8a36d0329/src/hooks/useArrowKeyFocusManager.ts#L71

> when the focusIndex == 0, focus currentLocation button, and the rest for the suggestion list.

When the focusIndex == 0 the currentlocationitem cannot be selected because the arrowkeyfocusmanger returns early if maxindex is less than 0, also the address component search results are 0 index based and the currentlocation item is not part of the search results. 

> maxIndex as suggestionCount + currentLocationItem

I also tried adding placeholder item to arrowKeyFocusManager by setting maxIndex to 1 but the maxIndex is reset to 0.

> The suggestion was to investigate if we can improve the useArrowKeyFocusManager if required.

I am not finding value in refactoring the useArrowKeyFocusManager and the currentLocation item component is only for the address search. The arrowKeyFocusManager is not designed to handle selection outside of search results and the limitation is also caused by the autoplacecomplete component which does not include currentLocation item as part of search result.

> I am not sure why we have removed the useArrowKeyFocusManager and changed to raw key press handling?

The useArrowKeyFocusManager can only handle search result focus and not current location item but the key press event can handle both, so I suggested using only keypress event to handle both.

In previous commit, I had used both useArrowKeyFocusManager for handling search result selection and key press event for handling currentlocation item selection.

@roryabraham I am open to suggestion what to ugo with, useArrowKeyFocusManager+ KeyPress or only KeyPress event.

jsdev2547 commented:
@abdulrahuman5196 The useArrowKeyFocusManager returns early if the maxIndex is less than 0 and resets index to -1 when no search results, which is why I had previously used arrowkeyfocusmanager alongside with keypress event to handle currentlocation item selection.
https://github.com/Expensify/App/blob/c848cb6bf29b600eecb8b338222fffe8a36d0329/src/hooks/useArrowKeyFocusManager.ts#L71

> when the focusIndex == 0, focus currentLocation button, and the rest for the suggestion list.

When the focusIndex == 0 the currentlocationitem cannot be selected because the arrowkeyfocusmanger returns early if maxindex is less than 0, also the address component search results are 0 index based and the currentlocation item is not part of the search results. 

> maxIndex as suggestionCount + currentLocationItem

I also tried adding placeholder item to arrowKeyFocusManager by setting maxIndex to 1 but the maxIndex is reset to 0.

> The suggestion was to investigate if we can improve the useArrowKeyFocusManager if required.

I am not finding value in refactoring the useArrowKeyFocusManager and the currentLocation item component is only for the address search. The arrowKeyFocusManager is not designed to handle selection outside of search results and the limitation is also caused by the autoplacecomplete component which does not include currentLocation item as part of search result.

> I am not sure why we have removed the useArrowKeyFocusManager and changed to raw key press handling?

The useArrowKeyFocusManager can only handle search result focus and not current location item but the key press event can handle both, so I suggested using only keypress event to handle both.

In previous commit, I had used both useArrowKeyFocusManager for handling search result selection and key press event for handling currentlocation item selection.

@roryabraham I am open to suggestion what to go with, useArrowKeyFocusManager+ KeyPress or only KeyPress event.

jsdev2547 commented:
@abdulrahuman5196 The useArrowKeyFocusManager returns early if the maxIndex is less than 0 and resets index to -1 when no search results, which is why I had previously used arrowkeyfocusmanager alongside with keypress event to handle currentlocation item selection.
https://github.com/Expensify/App/blob/c848cb6bf29b600eecb8b338222fffe8a36d0329/src/hooks/useArrowKeyFocusManager.ts#L71

> when the focusIndex == 0, focus currentLocation button, and the rest for the suggestion list.

When the focusIndex == 0 the currentlocationitem cannot be selected because the arrowkeyfocusmanger returns early if maxindex is less than 0, also the address component search results are 0 index based and the currentlocation item is not part of the search results. 

> maxIndex as suggestionCount + currentLocationItem

I also tried adding placeholder item to arrowKeyFocusManager by setting maxIndex to 1 but the maxIndex is reset to 0.

> The suggestion was to investigate if we can improve the useArrowKeyFocusManager if required.

I am not finding value in refactoring the useArrowKeyFocusManager and the currentLocation item component is only for the address search. The arrowKeyFocusManager is not designed to handle selection outside of search results and the limitation is also caused by the autoplacecomplete component which does not include currentLocation item as part of search result.

> I am not sure why we have removed the useArrowKeyFocusManager and changed to raw key press handling?

The useArrowKeyFocusManager can only handle search result focus and not current location item but the key press event can handle both, so I suggested using only keypress event to handle both.

In previous commit, I had used both useArrowKeyFocusManager for handling search result selection and key press event for handling currentlocation item selection.

@roryabraham I am open to suggestion what to go with, useArrowKeyFocusManager + KeyPress or only KeyPress event.

jsdev2547 commented:
@abdulrahuman5196 PR is awaiting review.

abdulrahuman5196 commented:
I will go through these callouts https://github.com/Expensify/App/pull/37558#issuecomment-2030213878. I am not sure if updating the arrowKeyFocusManager would be a complex task, but anyways I might need to dive deep on this.

So will provide updates in 2 days with the path forward.

jsdev2547 commented:
@abdulrahuman5196 I have merged main. I am still waiting on the path forward.

abdulrahuman5196 commented:
@jsdev2547 I made some test code changes and I don't see any issue with the `arrowKeyFocusManager`

This is the test code I have added in your PR and removed the `onKeyPress` you have.

```
let arrowKeyMaxIndex = maxIndex
    if (shouldShowCurrentLocationButton) {
        arrowKeyMaxIndex = 1
    }
    const [focusedIndex, setFocusedIndex] = useArrowKeyFocusManager({initialFocusedIndex: -1, maxIndex: arrowKeyMaxIndex - 1 , isActive: true});
    console.log("Abdul Test printing focused index = " + focusedIndex)
```

I am getting proper focusedIndex value, attaching the reference video. We just need to handle cases to show hover effect on the current location button when needed.



https://github.com/Expensify/App/assets/46707890/a89529df-8882-49f9-902b-4bcee5e97b3e
